### ISSUE TITLE
"The Following content is not available on this app." message shown in video/5 minutes of silence instead of a sound

### ISSUE THREAD

**Issue by Milkdrop** [TeamNewPipe/NewPipe#8713]
### Checklist

- [X] I am able to reproduce the bug with the [latest version](https://github.com/TeamNewPipe/NewPipe/releases/latest).
- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Affected version

0.23.1

### Steps to reproduce the bug

1. Go to penguinz0's "Biggest scandal I've ever seen" video (https://www.youtube.com/watch?v=7-BqJKE14Yo)
2. Play it at 360p resolution

### Expected behavior

The video should play normally

### Actual behavior

The message "The following content is not available on this app. Watch this content on the latest version on YouTube" appears.

### Screenshots/Screen recordings

<img alt="Screenshot_20220731-020841" src="https://user-images.githubusercontent.com/17289219/182005371-16943816-d1f1-42e0-82e2-42c666fa3587.png" height=600 />

### Affected Android/Custom ROM version

Android 11 / ArrowOS 11

### Affected device model

ASUS ZenFone Max Pro M1

**Comment by LuanRT** [comment]
I noticed this also happens in older versions of official clients, and it's quite random. So YouTube is probably rolling out some sort of update.

For those who haven't seen yet: https://support.google.com/youtube/answer/12318250?hl=en-GB

Similar issues:
https://github.com/iv-org/invidious/issues/3230

**Comment by AudricV** [comment]
Yes, that's random because when I try to play the video, it works for me. So seems to be a A/B test or a very slow rollout for me.

A workaround you can use for now is to block the `youtubei.googleapis.com` domain on NewPipe, which will prevent the extractor to use Android and iOS clients, and make only use of the desktop web client.

We should update the client version used to the latest one (`17.29.34` on Android and `17.29.3` on iOS currently) and see if the issue happens. If so, unfortunately, we would have to remove usage of other clients than the HTML5 ones or make better spoofing of the mobile clients.

We use right now the `17.10.35` version, as you can see [here](https://github.com/TeamNewPipe/NewPipeExtractor/blob/dev/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java#L178).

**Comment by coletdjnz** [comment]
Seems like [changing the app version may not suffice.](https://github.com/Tyrrrz/YoutubeExplode/issues/647#issuecomment-1204313293)

Since this is an A/B Innertube API test, I'm assuming it is tied to `vistorData`.

If sometime can get their hands on either:
- `visitorData` string (`X-Goog-Visitor-Id` header, or `visitorData` string in the `context` of an API request  or in the `responseContext` of an API response)
-  `VISITOR_INFO1_LIVE` cookie

from a request or session that experiences this it might be very helpful in getting a reproducible environment.

**Comment by unixfox** [comment]
@coletdjnz Could the API key be outdated too?

**Comment by AudricV** [comment]
According to my tests, the visitor data (at least in the header), user-agent, API key, `t` and `id` parameters of the player request does not have an effect on URLs: everything seems to be related to the protobuf body sent.

> Could the API key used in Newpipe be outdated too?

No, the API key used is still the same, as you can see on this capture made by mitmproxy:

<img src="https://user-images.githubusercontent.com/74829229/182879941-d938b6b5-2019-4baf-a452-50eeaf2866cf.png" width="500px" alt="mitmproxy_player_fallback_request">

**Comment by unixfox** [comment]
> If this is to be believed, we need to get our hands on a reproducible environment for us to test on.

I have a visitorData ID that was able to produce the issue, I've packed a bunch of info in that gist: https://gist.github.com/unixfox/a7ba15fa9b71ab302f7eebf7aa876563

Here is my dirty script for continuously checking the hash of a YouTube video: https://gist.github.com/unixfox/10a3f4e5a8499059ff18ad8d6b260585

I replicated the issue on a [Digitalocean](https://www.digitalocean.com/) droplet in UK location with IPv6 disabled and the DNS server set to `8.8.8.8` if one wants to have an environment to the one that I used. Outbound traffic is free on Digitalocean.

**Comment by unixfox** [comment]
Through trial and errors, I just found in the protobuf player request that the parameter needed to not have the error `The Following content is not available on this app.` is this one:
```
64: 29
```

In the protobuf decoded raw request (`cat player | protoc --decode_raw`):
```
1 {
  1 {
    16: 3
    17: "17.29.34"
    64: 29
  }
}
2: "c7NdXldReAg"
```

Here is another preview of the protobuf decoded raw request from this tool https://github.com/nevermoe/protobuf_decoder:
```json
{
    "01:00:embedded message": {
        "01:00:embedded message": {
            "16:02:Varint": 3, 
            "17:03:string": "17.29.34",
            "64:20:Varint": 29
        }
    }, 
    "02:01:string": "c7NdXldReAg"
}
```

The issue is that I've no idea what's the 64th element of the `context => client` object.

In summary:
- `1` is `context`
- `1 => 1` is `context => client`
- `1 => 1 => 16` is `context => client => "ANDROID"`
- `1 => 1 => 17` is `context => client => "17.29.34"`
- `2` is `videoId`

I've uploaded the protobuf file here: [protobuf.zip](https://github.com/TeamNewPipe/NewPipe/files/9276616/protobuf.zip)

I'll continue my research but if more people can help me in finding that parameter, that would be great.

**Comment by LuanRT** [comment]
> The issue is that I've no idea what's the name of the 64th element of the `context => client` object, so that we can use it in a JSON request.

I've seen seen this parameter multiple times before — isn't it the Android API level?

Ex:
```
1 {
  1 {
    12: "Google"
    13 {
      6: 0x45303130
    }
    16: 3
    17: "17.26.32"
    18: "Android"
    19: "6.0"
    21 {
      12: 0x53552d6e
    }
    37: 320
    38: 569
    39: 0x40000000
    40: 0x4063bbbc
    41: 2
    46: 1
    50: 222613015
    55: 320
    56: 569
    61: 3
    64: 23 // 23 is Android 6.0
    65: 0x3fc00000
    67: 18446744073709551436
    80: "America/Sao_Paulo"
  }
}
```

So, if all they're doing is checking the OS version then maybe we could use the [`osVersion`](https://github.com/LuanRT/YouTube.js/blob/60ff0513f1cad07fc71a487652d84adf8a4cc8b1/src/core/Session.ts#L28) param in JSON requests?

**Comment by unixfox** [comment]
> So, if all they're doing is checking the OS version

You are correct, I'm on Android 10 and the API level is 29.

> then maybe we could just use the [`osVersion`](https://github.com/LuanRT/YouTube.js/blob/60ff0513f1cad07fc71a487652d84adf8a4cc8b1/src/core/Session.ts#L28) param in JSON requests?
> 
> Edit: Could also be `osApiLevel`, if that even exists.

Just tested, both `osVersion` and `osApiLevel` doesn't work. I think it's more something like `platformXXX`, it's not `platformVersion` nor `platformApiLevel`.

**Comment by SamantazFox** [comment]
@unixfox have you tried `clientSDK` or `platformSDK`?

Btw, I've gathered my current knowledge of the protobuf to parameter name mapping here:
https://gist.github.com/SamantazFox/be4a0f224b179c544d7dac85768fdd12

**Comment by unixfox** [comment]
> @unixfox have you tried `clientSDK`, `platformSDK`, or even just `SDK`?
> 
> Btw, I've gathered my current knowledge of the protobuf to parameter name mapping here: [gist.github.com/SamantazFox/be4a0f224b179c544d7dac85768fdd12](https://gist.github.com/SamantazFox/be4a0f224b179c544d7dac85768fdd12)

Yes doesn't work. The issue is that the parameter name is case-sensitive... that makes everything much more much difficult

**Comment by Prototaxis** [comment]
@unixfox Give `androidSdkVersion` a try. I found it in the youtube app headers over here: https://github.com/thelazyitguy/AppHeaders/blob/36333bdff322f80a5895d7441f4cdf68637812fa/YouTube%2015.45.2/YTIClientInfo.h

**Comment by unixfox** [comment]
> @unixfox Give `androidSdkVersion` a try. I found it in the youtube app headers over here: [thelazyitguy/AppHeaders@`36333bd`/YouTube%2015.45.2/YTIClientInfo.h](https://github.com/thelazyitguy/AppHeaders/blob/36333bdff322f80a5895d7441f4cdf68637812fa/YouTube%2015.45.2/YTIClientInfo.h)

Thank you @Prototaxis! That's the correct parameter name! I don't get the error anymore.

**Comment by Prototaxis** [comment]
I'm glad it worked. Thanks @coletdjnz for forwarding me to this thread.

**Comment by Prototaxis** [comment]
In general, app headers are very useful for reverse engineering the protobuf. I have the comments layer mapped, browse and next partially, too. Let me know if you need help with those.

**Comment by SamantazFox** [comment]
> In general, app headers are very useful for reverse engineering the protobuf. I have the comments layer mapped, browse and next partially, too. Let me know if you need help with those.

yes, that would be extremely helpful :)

**Comment by 89z** [comment]
https://github.com/zerodytrash/YouTube-Internal-Clients/issues/3
